[ConstraintSolver] NFC: Increase default shrinking termination threshold

xedin opened the pull request:
We've found in practice that multiple different types of expressions
are still going to benefit from shrinking continuing even when it
couldn't simplify up to 10 sub-expressions.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
@swift-ci please smoke test and merge

xedin commented:
@swift-ci please smoke test OS X platform

xedin commented:
@swift-ci please smoke test OS X platform

xedin commented:
@swift-ci please smoke test

xedin commented:
@rudkx Looks like if I touch the threshold on master (even slightly) some of our perf tests fail because they can no longer fit into neither polynomial nor exponential model :(

xedin commented:
I've ran a couple of experiments using `validation-test/Sema/type_checker_perf/fast/rdar19915443.swift.gyb` and I'm not quiet sure why can't the model be determined e.g. `--begin 6 --end 20 --step 1` produces `[823, 850, 886, 922, 958, 994, 1030, 1066, 1102, 1138, 1174, 1210, 1246, 1282]` which is a pretty even step. 

@graydon Could you please clarify a bit about how fitting works here?

felix91gr commented:
> Looks like if I touch the threshold on master (even slightly) some of our perf tests fail because they can no longer fit into neither polynomial nor exponential model :(

That clearly shouldn’t happen... either the tests aren’t quite right, or there’s something extraneous affecting the result. Are they giving failing messages? Maybe they are getting timeout-ed.

> how fitting works here?

Afaik we are doing linear regression over a list of samples (Y axis) + a list of values of the independent variable (X axis). For each model O(f(x)) to be fit, the samples are transformed before applying the regression. That transformation is the inverse of the model (ie **log_2(x)** if we’re fitting **2^x**). The result is a regression that fits best the more the model **f(x)** is like the measured model.

That is as far as I understood the model regression script to be. Someone re-wrote it recently, and I’m not 100% sure of how it is now. Maybe they introduced a bug, but that would be surprising. That’s how I’m almost sure it used to work for the past year until recently, anyway.

graydon commented:
The problem is you've got a linear model with a strong deviation from it on the first point (that is: every gap after the first is 36, but the first gap is 27). Push the starting point over by one.

As for how it works: the previous curve-fitting worked by transforming the data to log-log space and trying to do linear regression, as @felix91gr describes, which it then interpreted the slope as an exponent. Unfortunately that interpretation is wrong when the function doesn't pass through the origin, which many of our data don't. So we'd get "good fits" that were wrong; meaningless.

So I rewrote it to do a numerical nonlinear regression: an iterative simplex-walking algorithm to attempt to minimize the least-squares difference between polynomial and/or exponential function parameters and the input data, in non-transformed (linear) space. This is an overall somewhat less stable / reliable process -- it fails-to-fit more often -- but it also fits more types of data correctly, rather than coming up with meaningless "good" fits in log-space.

xedin commented:
@graydon Thanks for explanation! Is there anything which could be done about that without modifying the test itself?

felix91gr commented:
> Unfortunately that interpretation is wrong when the function doesn't pass through the origin, which many of our data don't. So we'd get "good fits" that were wrong; meaningless.

Mm, indeed. I didn´t even notice that. You're totally right.

> So I rewrote it to do a numerical nonlinear regression: (...). This is an overall somewhat less stable / reliable process -- it fails-to-fit more often -- but it also fits more types of data correctly, rather than coming up with meaningless "good" fits in log-space.

Well, it's better than what we had before, that's for sure. But the stability remains an open issue. Have you looked at SciPy's [`curve_fit`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.curve_fit.html)? Maybe that's enough, and given it's SciPy, we've got a good chance on it being stable with more input.

xedin commented:
@graydon Just want to note that I've tried locally with more steps and different start `--begin 6 --end 20 --step 1` which produces `[823, 850, 886, 922, 958, 994, 1030, 1066, 1102, 1138, 1174, 1210, 1246, 1282]` but test still fails saying that it can't fit, does that make sense?

graydon commented:
@xedin Yes, what I'm saying is: try with `--begin 7` or `--begin 8`. The first point is off the line, messing up the linear fit.

@felix91gr Oh, there are plenty of good nonlinear optimization algorithms! I'd far prefer to use numpy's more modern / mature / less-buggy implementations here, unfortunately we don't want to add extra python dependencies to our testsuite. That's why I implemented it myself (and why I used a relatively old, low-quality algorithm -- only one I had a chance of implementing myself in reasonable time / reasonably-low level of bugginess).

More generally: I'm happy to continue refining the algorithm here (and/or replacing it if we're willing to take a dependency). I already put in a check for perfect linear fits to avoid this sort of thing; I suspect it's worth expanding that to run a best-linear fit pass, before trying higher-degree polynomials. IOW I think the problem you're seeing here is that the simplex algorithm gets stuck alternating between adjusting the linear factor and the polynomial exponent. It should probably run a separate fit on _just_ a linear model to see if that fits overall-better.

xedin commented:
@graydon Thank, let me see if I can re-adjust the steps etc. to make it all work again. I think long term we have to do something about the actual scopes counter too to make it more _real_...

felix91gr commented:
@graydon I see. I'd rather add the dependency: it's better to have a well-tested testing tool, so that the community can focus on fixing the compiler, instead of having to fix both the compiler and the tools.

But for the moment, lemme know if I can help you with what we have. Do you have a to-do list? Maybe this can be filed as a bug, and I can implement it with you giving me feedback :)

huonw commented:
>  I'd rather add the dependency

So would we all (there's no technical reason to implement curve fitting ourselves), but it is *very* annoying to have to get the extra dependency onto the testing bots. You can definitely argue that this is an unfortunate setup, but it's the current reality and something can't be fixed in scale-test (or even in this repo).

felix91gr commented:
> it is _very_ annoying to have to get the extra dependency onto the testing bots.

Ah, it's worse than I imagined. I thought only of the users of the testing suite, which would have little trouble installing the dependencies. But the issue is with the bots... now I get it.

I'm gonna look up curve fitting libraries, maybe there's one that we can use without runtime and with the full source on the repo. Such a thing probably doesn't exist, but I'm gonna take a look anyway :thinking:

xedin committed:
[ConstraintSolver] NFC: Increase default shrinking termination threshold

We've found in practice that multiple different types of expressions
are still going to benefit from shrinking continuing even when it
couldn't simplify up to 10 sub-expressions.

xedin commented:
@swift-ci please smoke test
